### DISCUSSION TITLE
Error: Text content does not match server-rendered HTML when updating a server rendered list of items

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#43317] [Help]
I have a page that shows a video and a list of comments for that video. I chose server side rendering because comments should be included on every initial page load and I wanted the streamlined simplicity of writing a single getServerSideProps rather than the overhead of writing an API route and then making a client side fetch. Later I realized that users should be able to add new comments to the list by submitting a new comment form that exists on the same page. I wrote an API route to create the comment and then when the form is submitted I call a form handler that makes a POST call to the API route to create the new comment. Next I was confronted with the problem of refreshing the list for the user so they can see their new comment. Since my page is server side rendered, I thought I'd be clever by calling router.reload() and refreshing the page after the HTTP POST succeeds. 

It actually works in that after comment creation the page reloads and I see the new comment... but only for the first 3 (sometimes 4) reloads. After the 3rd (or 4th) comment posted (and reload called), I get the following error: Error: Text content does not match server-rendered HTML. I researched this and learned that this error occurs when the the React tree during server side rendering does not match the tree in the browser. I am a little surprised that my use case triggers this error as I would have figured that a router.reload() would grab a fresh server rendered page. I am also surprised that this error occurs somewhat unpredictably and does not occur in the first couple reloads. 

1. Does the use case of initially displaying 10 items of X and having a form add a new X to the list to show a larger list of 11 items of X rule out the ability to use server-side rendering? This would mean that SSR cannot be used to give initial lists to browsers if those lists were to ever change from browser interaction. In other words, the rule could be simplified as, "if a list is going to be updated, it must be client rendered"
2. Am I understanding this error correctly?

**Comment by icyJoseph** [comment]
## Refreshing GSSP props

>  I wrote an API route to create the comment and then when the form is submitted I call a form handler that makes a POST call to the API route to create the new comment. Next I was confronted with the problem of refreshing the list for the user so they can see their new comment. Since my page is server side rendered, I thought I'd be clever by calling router.reload() and refreshing the page after the HTTP POST succeeds.

I stopped reading there because there's an actual solution for this, that doesn't require `reload`.

Allow me to drop an article which expands on it, https://www.joshwcomeau.com/nextjs/refreshing-server-side-props/:

```js
import { useRouter } from 'next/router';
function SomePage(props) {
  const router = useRouter();
  // Call this function whenever you want to
  // refresh props!
  const refreshData = () => {
    router.replace(router.asPath);
  }
}
export async function getServerSideProps(context) {
  // Database logic here
}
```

So calling, `replace`, on the same `asPath`, keeps the queries, keeps state and what not, and simply, executes GSSP server side, and uses the response as props to your component, all without reloading the page!

## Onto your issue

I imagine the issue you are seeing is caused by time differences. Are there any timestamps attached to the comments? Like `a minute ago` or `posted 10 hours ago` kind of things? 

Say during SSR your logic sees that a comment happened 59 seconds ago, and it renders `less than a minute ago`, but when the client receives the HTML (because you are doing `reload`), and hydrates it, now it has been 60 seconds ago, so you render `a minute ago` and BAM HTML has diverged...

## Solution

For OP's particular case, the following is enough:

```jsx
<span suppressHydrationWarning>{formattedTimestamp}</span>
```

**Reply by icyJoseph** [reply]
@sharpsteelsoftware1 Because it is only around the "edges" of a time change (from seconds to minutes, to hours, etc), it'll be really hard to try and find a good balanced solution, given that the "edge" can happen while the data is traveling from server to client, so it'd say do as the React docs indicate, and use https://reactjs.org/docs/dom-elements.html#suppresshydrationwarning:

> If you use server-side React rendering, normally there is a warning when the server and the client render different content. However, in some rare cases, it is very hard or impossible to guarantee an exact match. For example, timestamps are expected to differ on the server and on the client.
> 
> If you set suppressHydrationWarning to true, React will not warn you about mismatches in the attributes and the content of that element. It only works one level deep, and is intended to be used as **_an escape hatch._** **_Don’t overuse it_**. You can read more about hydration in the [ReactDOM.hydrateRoot() documentation](https://reactjs.org/docs/react-dom-client.html#hydrateroot).

Use it like so:

```jsx
<span suppressHydrationWarning>{formattedTimestamp}</span>
```

**Reply by ghost** [reply]
Thank you very much. I love NextJS and your support is phenomenal, I will be bringing more NextJS development into the industry!

**Reply by icyJoseph** [reply]
Let us know if it works, and don't forget to mark as answered :)

**Comment by ghost** [comment]
Yes that is it! Good catch!! Hmm. So what does that mean if we have timestamps and are using '59 seconds ago' etc?
